IdentityCore/src/MSIDJsonSerializableFactory.m

```diff
@@ -0,0 +1,99 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDJsonSerializableFactory.h"
+#import "MSIDJsonSerializable.h"
+
+static NSMutableDictionary<NSString *, Class<MSIDJsonSerializable>> *s_container = nil;
+
+@implementation MSIDJsonSerializableFactory
+
++ (void)registerClass:(Class<MSIDJsonSerializable>)class forKey:(NSString *)key
+{
+    if (!class || !key) return;
+    if (![key isKindOfClass:NSString.class]) return;
+    if (![class conformsToProtocol:@protocol(MSIDJsonSerializable)]) return;
+    
+    @synchronized(self)
+    {
+        static dispatch_once_t once;
+        dispatch_once(&once, ^{
+            s_container = [NSMutableDictionary new];
+        });
+        
+        s_container[key] = class;
+    }
+}
+
++ (void)unregisterAll
+{
+    @synchronized(self)
+    {
+        [s_container removeAllObjects];
+    }
+}
+
++ (id<MSIDJsonSerializable>)createFromJSONDictionary:(NSDictionary *)json
+                                        classTypeKey:(NSString *)classTypeKey
+                                               error:(NSError **)error
+{
+    if (![json msidAssertType:NSString.class ofKey:classTypeKey required:YES error:error]) return nil;
+    NSString *classTypeValue = json[classTypeKey];
+    
+    return [self createFromJSONDictionary:json containerKey:classTypeValue error:error];
+}
+
++ (id<MSIDJsonSerializable>)createFromJSONDictionary:(NSDictionary *)json
+                                      classTypeValue:(NSString *)classTypeValue
+                                               error:(NSError **)error
+{
+    return [self createFromJSONDictionary:json containerKey:classTypeValue error:error];
+}
+
+#pragma mark - Private
+
++ (id<MSIDJsonSerializable>)createFromJSONDictionary:(NSDictionary *)json
+                                        containerKey:(NSString *)containerKey
+                                               error:(NSError **)error
+{
+    Class class = (Class<MSIDJsonSerializable>)s_container[containerKey];
+    
+    if (!class)
+    {
+        NSString *errorMessage = [NSString stringWithFormat:@"Failed to create object from json, class: %@ wasn't registered in factory under %@ key.", class, containerKey];
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInvalidDeveloperParameter,
+                                     errorMessage,
+                                     nil, nil, nil, nil, nil);
+        }
+        
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, nil, @"%@", errorMessage);
+        return nil;
+    }
+    
+    return [[(Class)class alloc] initWithJSONDictionary:json error:error];
```

Comment:
Let's also double check we get back expected class

Reply:
done